server/WFCORE-6750-unstable-api-annotation-scanner.adoc

```diff
@@ -0,0 +1,243 @@
+
+=== QE Contacts
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+* [x] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+
+* WildFly
+** the wildfly-galleon-pack and wildfly-ee-feature pack feature packs will scan their classpaths and provide the indices of API elements annotated with `io.smallrye.common.annotation.Experimental` and `org.hibernate.Incubating`.
+* WildFly Core
+** scans the applications when they are deployed, looking for usage of API elements that exist in the index.
+* https://github.com/wildfly/unstable-api-annotation-utils[wildfly/unstable-api-annotation-utils] - contains a library providing:
+** A Maven plugin which can be used by the feature packs to do the scanning
+** A highly optimised bytecode scanner and lookup mechanism to find usage in user applications of API elements that appear in the indices created by the feature packs
+
+
+=== Other Interested Projects
+
+Feature packs external to WildFly, as mentioned in the 'Future Work' section.
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+* Note this is at preview level only!
+* There is a maven plugin that can be used in the feature pack builds.
+** It will scan for annotations indicating unstable API, and add the following to the `content/` directory of the  `{indexModule}` module (the `content/` directory is defined as a resource root:
+*** a file containing the indices for annotations concerning the feature pack.
+*** The name of the file will be added to a file called `index.txt` in the `content/` folder of the `{indexModule}` module. This `index.txt` is provided as the main lookup mechanism for the files mentioned in the above point.
+** The maven plugin scans the full classpath of the feature pack looking for dependencies to index.
+** For this iteration we will scan
+*** the `wildfly-ee-feature-pack` for the `org.hibernate.Incubating` annotation.
+*** the `wildfly-galleon-pack` for the `io.smallrye.common.annotation.Experimental` annotation.
+*** It also contains a simple configuration mechanism to narrow down which classpath entries to index (e.g when scanning for `io.smallrye.common.annotation.Experimental` we only check the SmallRye libraries on the classpath)
+* In WildFly Core
+** deployment unit processors will be added to:
+*** Load the indices listed in the `content/index.html` file in the `{indexModule}` module.
+*** Use the wildfly/unstable-api-annotation-utils library to scan the bytecode of every class in the user's deployed application and cross-reference against the loaded indices.
```

Comment:
This should be provided by a module that is an optional dependency of whatever uses this. That module has property jboss-stability-level=preview.

That likely means the index files should be in a different module from this lib, one that's at default stability, so the plugin execution that adds them doesn't have to know whether the FP build is configured to allow preview stability modules in the FP.

Reply:
I misspoke -- it's jboss.stability=preview, for example

https://github.com/wildfly-extras/mvc-krazo-feature-pack/blob/main/galleon-shared/src/main/resources/modules/system/layers/base/jakarta/mvc/api/main/module.xml

For a brief discussion of this see 'JBoss Modules modules stability level' in https://docs.wildfly.org/wildfly-proposals/wf-galleon/WFLY-19021-Stability_In_Provisioning.html#requirements-at-feature-pack-build-time